Refactoring base vel/acc IK codes with a base class

robodreamer opened the pull request:
- all the common functions are moved to base class now
- updated velocity and acceleration IK codes so that
  they use the functions in the base class

This update was written based on Matt's initial attempt. I fixed a bug in his version and made acceleration IK use base class functions as well.

Commit:
Create a base SNS IK class

- all the common functions are moved to base class now
- updated velocity and acceleration IK codes so that
  they use the functions in the base class

Commit:
Addressed Matt's comment

Moved Eigen-based formatting to base class.

robodreamer commented:
I have updated the code based on Matt's comments.

Commit:
Updated author names

Commit:
Remove some incorrect descriptions

Removed the description in each file that was carried
from an older file that Matt started with.

Commit:
Remove duplicate code in Matlab functions

In velocity/acceleration IK with configuration task,
we now reuse the base IK function for primary task.

IanTheEngineer commented on sns_ik_lib/include/sns_ik/sns_ik_base.hpp:
```diff
@@ -0,0 +1,248 @@
+   * @param qUpp: upper bound on the velocity/acceleration of each joint
+   * @return: true iff successful
+   */
+  bool setBnd(const Eigen::ArrayXd& qLow, const Eigen::ArrayXd& qUpp);
```
This strikes me as an unnecessary abbreviation. I'd prefer to see it named `setBounds` or even `setJointConfigBounds`

IanTheEngineer commented on sns_ik_lib/include/sns_ik/sns_ik_base.hpp:
```diff
@@ -0,0 +1,248 @@
+  /*
+   * @return: reference to the lower bound in configuration space (pos / vel / acc)
+   */
+  const Eigen::ArrayXd& getLow() const { return qLow_; };
```
Again, the more descriptive names the better, ie. `getLowerBounds`

IanTheEngineer commented on sns_ik_lib/include/sns_ik/sns_ik_base.hpp:
```diff
@@ -0,0 +1,248 @@
+  /*
+   * @return: number of joints
+   */
+  int nJnt() const { return nJnt_; }
```
I'd follow KDL's lead and name this [`getNrOfJoints()`](https://github.com/orocos/orocos_kinematics_dynamics/blob/4c09fb9849da8952107a5efe7aa6b2ca58c68d0e/orocos_kdl/src/chain.hpp#L71)

IanTheEngineer replied:
The return value should be of type `unsigned int` or a `size_t`

robodreamer replied:
Good suggestion.

IanTheEngineer commented on sns_ik_lib/include/sns_ik/sns_ik_base.hpp:
```diff
@@ -0,0 +1,248 @@
+   * @return: true iff qLow <= q <= qUpp
+   *          if q.empty() return false
+   */
+  bool checkBnd(const Eigen::VectorXd& q);
```
`checkBounds`

IanTheEngineer commented on sns_ik_lib/include/sns_ik/sns_ik_base.hpp:
```diff
@@ -0,0 +1,248 @@
+  /*
+   * @return: reference to the lower bound in configuration space (pos / vel / acc)
+   */
+  const Eigen::ArrayXd& getUpp() const { return qUpp_; };
```
`getUpperBounds`

IanTheEngineer commented on sns_ik_lib/include/sns_ik/sns_ik_base.hpp:
```diff
@@ -0,0 +1,248 @@
+  /*
+   * @return: rank of the matrix that is currently set in the linear solver
+   */
+  int getLinSolverRank() const { return linSolver_.rank(); }
```
Can this solver return a negative rank?

robodreamer replied:
No it shouldn't. I will change it to size_t as well.

IanTheEngineer replied:
in this context, i think `unsigned int` makes more sense, as it does not represent the size of a std container

IanTheEngineer commented on sns_ik_lib/src/sns_acc_ik_base.cpp:
```diff
@@ -226,40 +168,36 @@ SnsAccIkBase::ExitCode SnsAccIkBase::solve(const Eigen::MatrixXd& J, const Eigen
-      ddqNull(jntIdx) = ddqUpp_(jntIdx);
-    } else if ((*ddq)(jntIdx) < ddqLow_(jntIdx)) {
-      ddqNull(jntIdx) = ddqLow_(jntIdx);
+    if ((*ddq)(jntIdx) > (getUpp())(jntIdx)) {
```
A prime example of where the naming changes that I outlined above would make this code easier to read.

robodreamer replied:
I agree.

IanTheEngineer reviewed (changes requested):
These changes are great! I just have recommendations on naming conventions

Commit:
Addressed Ian's comments

- changed some function names to more descriptive ones.
- change the type of indices to unsigned integer type (size_t)

IanTheEngineer reviewed (approved):
Perfect, just modify the last return type from `size_t` to `unsigned int` and i'll squash merge.

Commit:
change return type from size_t to uint

Commit:
remove white space

IanTheEngineer commented on sns_ik_lib/include/sns_ik/sns_ik_base.hpp:
```diff
@@ -149,7 +149,7 @@ class SnsIkBase {
    * @return: rank of the matrix that is currently set in the linear solver
    */
-  size_t getLinSolverRank() const { return linSolver_.rank(); }
+  uint getLinSolverRank() const { return linSolver_.rank(); }
```
`uint` actually isn't a defined type in C++. This must be `unsigned int`

IanTheEngineer commented on sns_ik_lib/include/sns_ik/sns_ik_base.hpp:
```diff
@@ -0,0 +1,248 @@
+  /*
+   * @return: rank of the matrix that is currently set in the linear solver
+   */
+  uint getLinSolverRank() const { return linSolver_.rank(); }
```
This needs to be `unsigned int` as `uint` isn't guaranteed to be defined.

Commit:
modified types of rank variables
